### PR TITLE
✨ Label interruptible nodes

### PR DESCRIPTION
**Pull request by alexander-demichev**
**What this PR does / why we need it**:

Transfer machine labels to the node. The use-case is described here https://github.com/kubernetes-sigs/cluster-api/issues/3504. 

The transfer of machine labels to the node is similar to setting `Noderef`. Important thing to mention is that only labels with `"cluster.x-k8s.io"` prefix are transferred.

This PR should help with implementing the termination handler for spot instances - https://github.com/kubernetes-sigs/cluster-api/pull/3528

Happy to hear what others think about this change.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3504

### CODE REVIEW

File: controllers/machine_controller_nodelabels.go
Comment (JoelSpeed) [inline controllers/machine_controller_nodelabels.go:2]:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

File: controllers/machine_controller_nodelabels.go
Comment (JoelSpeed) [inline controllers/machine_controller_nodelabels.go:66]: Is there a constant that can be used for this somewhere already?

**Review by JoelSpeed — COMMENTED** [review]
Looks good, couple of small comments but I have nothing major to add

**Comment by enxebre** [comment]
is this intending to supersede https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1876?
Overall seems reasonable to me.

**Comment by ncdc** [comment]
We have had previous requests about setting labels on nodes - see #458 and #493. The earlier consensus was that Cluster API will not reconcile node labels, but the kubeadm bootstrapper can set the initial values for a node's labels using `KubeadmConfig.spec.{initConfiguration,joinConfiguration}.nodeRegistration.kubeletExtraArgs["node-labels"]`. Continuing maintenance of a node's labels is better left to some other actor.

I do think it could be useful to adjust `Machine.spec` to add a field such as `initialNodeLabels` and then update the contract for bootstrap providers to use the values in that field to set the node's initial labels.

We will need to have further discussion on this before we can proceed, given the previous decision not to do this.

/hold

cc @kubernetes-sigs/cluster-api-maintainers

**Comment by alexander-demichev** [comment]
@ncdc Will list of labels be additive if we use kubeadm boostraper?

**Comment by ncdc** [comment]
@alexander-demichev they're the initial set of labels. See https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ (search for `--node-labels`) for more information.

**Comment by enxebre** [comment]
@ncdc @alexander-demichev is there a way to intrinsically infer a given machine is for preemptible so the node can be labeled based on this rather than generically syncing labels?
E.g Couldn't this be done by each specific provider implementation when they see their preemptible API fields set?

**Comment by ncdc** [comment]
@enxebre ah ok now I see what the desired behavior is here around labels & preemptible/spot instances. If we're looking for a behavior that is consistent across infrastructure providers, I'd say it's most appropriate to implement that behavior in core Cluster API, or at the very least, define the contract portion in the core.

A behavior such as "if a Machine is preemptible, make sure its Node has the label `cluster.x-k8s.io/interruptible-instance`" could be implemented
1. as one of the initial node labels set by the bootstrapper when registering the node
1. via reconciliation in the core Cluster API machine controller
1. via something outside of / on top of Cluster API

Option 1 (bootstrapper) requires every Cluster API bootstrap provider to support setting a node's initial labels. This seems reasonable to me. One downside is that if the interruptible label were somehow removed from the Node, nothing would add it back.

Option 2 (reconciliation) should probably only support syncing a very specific set of labels to a Node, namely this interruptible label for the time being. I wouldn't want to open this up to user-supplied labels, for the reasons previously stated in the other issues I linked above. But I do think it makes sense to sync labels defined by Cluster API that influence the behavior of the overall system. One downside to this option is that there would be a short period of time during which the interruptible label would be absent from the new Node. One upside is that the interruptible label would be re-added to the Node if it were somehow removed.

Option 3 (some external entity) I mentioned for completeness, but I don't think it's appropriate because Cluster API should be or is knowledgable that this is an interruptible machine.


For this specific use case -- ensuring a specific interruptible label is present on a Node -- I think it's appropriate to do both 1 and 2 together. 1 is already available in the kubeadm bootstrap provider and may be available in others. I do think a future API improvement could be what I mentioned in a previous comment - adding an `initialNodeLabels` spec field to Machine and having bootstrap providers use that. Option 2 would be new, and restricted solely to syncing this new interruptible label.


I also wonder if it would make sense to consider adding another new spec field to Machine - `interruptible` (name TBD) - to indicate the user understands this could be deleted by the cloud provider. This would simplify the burden on the user: they'd set `spec.interruptible`, and the machine reconciler would take care of setting the appropriate interruptible label (in other words, the user wouldn't have to spell out the full key name for the label and hope they don't have a typo). Adding a new `interruptible` field to the Machine spec has an added benefit for infrastructure providers - they could validate the corresponding InfraMachine (AWSMachine/AzureMachine/GCPMachine/etc) is configured with the interruptible field(s) appropriately. This idea could be implemented later, however - it's not required for the label syncing to function as desired.

**Comment by enxebre** [comment]
This all make sense to me @ncdc. As a user to define an interruptible machine you just need to set it in one place today, the infra template. I think we should aim to keep it that way so there's no room for partial supported scenarios. To that end I wonder if it'd make sense to let `interruptible` to be in the status rather than the spec (so we own it, not the user) and let option 2 to reconcile the node label based on this field being populated.

**Comment by ncdc** [comment]
> I wonder if it'd make sense to let interruptible to be in the status rather than the spec (so we own it, not the user) and let option 2 to reconcile the node label based on this field being populated.

I like this idea. As I was writing my spec.interruptible suggestion, I was wondering if it would be an unnecessary extra step. Brainstorming on flow:

1. User creates InfraMachine with whatever spec field(s) are required for that provider to indicate it's interruptible
1. Infra provider sets InfraMachine.status.interruptible=true
1. Machine controller syncs InfraMachine.status.interruptible to Machine.status.interruptible
1. (optional) Bootstrap provider somehow waits for the above sync so it can use Machine.status.interruptible when creating the bootstrap data secret (this needs more thinking)
1. Machine controller ensures the interruptible label is always present on the Node if Machine.status.interruptible is true

**Comment by CecileRobertMichon** [comment]
If we do move forward with this approach, can we also make sure that the same can be applied to MachinePools? I think the approach @ncdc describes above should work at first glance but I want to make sure we explicitly include it in the design (and even maybe implementation if it's straight forward).

**Comment by vincepri** [comment]
I'd hold off on adding more status syncing between infrastructure provider and the machine controller. From a high level perspective, Cluster API itself doesn't know that a Machine is preemptible.

The approach described above could potentially cause a dependency loop between the controllers. One alternative approach, which should be discussed in a proposal, is to add a condition to the InfrastructureMachine that communicates "ready to receive bootstrap data". Bootstrap providers would have to wait for this condition to appear before generating the bootstrap configuration and continue execution.

This would be a breaking change, which definitely requires some more thinking / design proposal. I'd suggest to open an RFE issue and gather use cases and proposals.

How does that sound?

/milestone v0.4.0

**Comment by enxebre** [comment]
>One alternative approach, which should be discussed in a proposal, is to add a condition to the InfrastructureMachine that communicates "ready to receive bootstrap data"...

I see this as something orthogonal and just a nice to have for the interruptible instances story. To cover the interruptible instances story end to end as a first iteration I see no concerns to proceed as with 1, 2, 3 and 5.
In parallel we can have a separate wider discussion for 4 and other use cases. wdyt?

**Comment by vincepri** [comment]
> In parallel we can have a separate wider discussion for 4 and other use cases. wdyt?

If the only goal is to update node labels after the Node is up, I think that could be doable. That said, adding a new synced status field goes somewhat against our different goal to move completely to conditions, and remove the current use of boolean fields.

**Comment by vincepri** [comment]
Is this something you'd like to do in v0.3.x, or can we wait until v0.4.0 and have an RFE / small proposal in an issue?

**Comment by ncdc** [comment]
I'd like to revise my previous suggestion and remove the part about adding interruptible to machine.status and syncing it from the infra machine to the machine. We can still achieve what we need by having the machine controller look at the infra machine's status (for interruptible), and avoid syncing that information to the machine.

> remove the current use of boolean fields.

Have we stated as a goal to never use any more booleans in status? Something like status.interruptible=true/false feels different than a condition, as it's data that's derived from the spec and isn't something that will ever change for the life of the infra machine. It seems ok to have a status boolean for this?

re timing, I propose we do this once we open main for new features & breaking changes for v0.4.0. Do you want a CAEP for this (contract change for infra providers to add status.interruptible (name/location TBD), machine controller updated to sync just this one label to interruptible nodes), or can we repurpose #3504 for the details?

**Comment by alexander-demichev** [comment]
@ncdc I think changing the description of https://github.com/kubernetes-sigs/cluster-api/issues/3504 and giving more context should be fine.

**Comment by alexander-demichev** [comment]
@ncdc @vincepri PTAL https://github.com/kubernetes-sigs/cluster-api/issues/3504 and let's move the discussion there.

**Comment by ncdc** [comment]
@alexander-demichev I just tagged #3817 for merge. Would you be willing to update this PR to implement the interruptible label syncing?

**Comment by alexander-demichev** [comment]
@ncdc I updated the PR, please take a look

File: controllers/machine_controller_node_labels.go
Comment (ncdc) [inline controllers/machine_controller_node_labels.go:47]: Please call `external.Get` to retrieve the infra object - that will avoid the race condition I mentioned above.

**Review by ncdc — CHANGES_REQUESTED** [review]
I am in meetings and need more time to review, but this was the foremost thing that stood out that needs some rework - thanks for doing this!

Could you please also retitle the PR and update the top description to reflect what it's doing now? Thanks!

File: controllers/machine_controller_node_labels_test.go
Comment (ncdc) [inline controllers/machine_controller_node_labels_test.go:39]: Any chance you could convert this to a test that uses the envtest available for the suite, instead of the fake client? You'd essentially create a Machine, InfraMachine, Node with all the right values filled in, then do an Eventually block to check to see if the Node's interruptible label is set.

File: controllers/machine_controller_phases.go
Comment (ncdc) [inline controllers/machine_controller_phases.go:258]: I think you'll end up reverting all the changes in this file (re race condition)

**Comment by alexander-demichev** [comment]
@ncdc Hi, I tried to address all comments, PTAL

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:44]: Can we merge these two together?

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:68]:
```suggestion
	if err := r.setInterruptibleNodeLabel(ctx, remoteClient, machine.Status.NodeRef.Name); err != nil {
```

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:57]:
```suggestion
```

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:72]:
```suggestion
	logger.V(3).Info("Set interruptible label to Machine's Node", "nodename", machine.Status.NodeRef.Name)
```

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:82]:
```suggestion
	if err := remoteClient.Get(ctx, client.ObjectKey{Name: nodeName}, node); err != nil {
```

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:102]:
```suggestion
	return patchHelper.Patch(ctx, node)
}
```

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:85]: We should check here if the label is already set, and return early if so.

**Review by vincepri — CHANGES_REQUESTED** [review]

File: controllers/machine_controller_node_labels.go
Comment (ncdc) [inline controllers/machine_controller_node_labels.go:51]: @vincepri what do you think about logging here and returning early instead of returning an error? If we return an error, it means we will start exponential backoff retries. I assume that unless the infra machine changes, continuing to retry the same object will yield the same failure each time. And, given we are already watching for infra machine changes, the machine will get requeued whenever the infra machine is updated.

File: controllers/machine_controller_node_labels.go
Comment (ncdc) [inline controllers/machine_controller_node_labels.go:34]: nit: consider moving this down to immediately before you first use the logger

File: controllers/machine_controller_node_labels.go
Comment (ncdc) [inline controllers/machine_controller_node_labels.go:79]: Can you move this down so we only instantiate if we know we need to patch?

File: controllers/machine_controller_node_labels.go
Comment (ncdc) [reply controllers/machine_controller_node_labels.go:78]: I think this is covered by L88-90?

File: controllers/machine_controller_node_labels_test.go
Comment (ncdc) [inline controllers/machine_controller_node_labels_test.go:92]: Please add a deferred call to `testEnv.Cleanup`, passing everything created so it can be deleted. You can find examples in other test files.

File: controllers/machine_controller_node_labels_test.go
Comment (ncdc) [inline controllers/machine_controller_node_labels_test.go:106]:
```suggestion
		err := testEnv.Get(ctx, client.ObjectKey{Name: node.Name}, updatedNode)
```

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [reply controllers/machine_controller_node_labels.go:51]: We definitely can. That said, these errors are uncommon code paths where we shouldn't get into unless something really wrong happened. Most of the time, functions like NestedBool are going to return regardless if the value exists or not, and in this case we use the default value for the boolean if the json path (status.interruptible) doesn't exists. The only way for this call to fail is if the `infra.Object` isn't well formed, which is unlikely in most cases, and it would probably hint to data corruption

Comment (ncdc) [reply controllers/machine_controller_node_labels.go:51]: Exactly, and if it's malformed data, exponential backoff isn't going to be useful here.

Comment (alexander-demichev) [reply controllers/machine_controller_node_labels.go:51]: That makes sense, thank you.

File: controllers/machine_controller_node_labels_test.go
Comment (ncdc) [inline controllers/machine_controller_node_labels_test.go:115]: nit: this line isn't necessary because it's already done in https://github.com/kubernetes-sigs/cluster-api/blob/ace4e245d02d6c4a73ba91534858e9c9a684a169/test/helpers/envtest.go#L76-L84.

**Comment by vincepri** [comment]
Can we rename this PR to better capture that it only supports interruptible label?

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:45]:
```suggestion
	log := ctrl.LoggerFrom(ctx)
```
To be consistent with the rest of the codebase

File: controllers/machine_controller_node_labels.go
Comment (vincepri) [inline controllers/machine_controller_node_labels.go:74]:
```suggestion
```

**Review by vincepri — COMMENTED** [review]
One minor comment, otherwise LGTM

**Comment by alexander-demichev** [comment]
@ncdc @vincepri All fixed, thanks for reviews!

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
